hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java

```diff
@@ -457,6 +461,37 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile, Hood
   private JavaRDD<WriteStatus> upsertRecordsInternal(JavaRDD<HoodieRecord<T>> preppedRecords, String commitTime,
       HoodieTable<T> hoodieTable, final boolean isUpsert) {
 
+    if (getConfig().getSchemaCheck()) {
+      // Ensure that the current writerSchema is compatible with the latest schema of this
+      // dataset.
+      // When inserting/updating data, we read records using the schema saved in the
+      // data/log files and convert them to the GenericRecords with writerSchema.
+      // Hence, we need to ensure that this conversion can take place without errors.
+      try {
+        SchemaUtil schemaUtil = new SchemaUtil(hoodieTable.getMetaClient());
+        MessageType savedParquetSchema = schemaUtil.getDataSchema();
+        Schema savedSchema = schemaUtil.convertParquetSchemaToAvro(savedParquetSchema);
+        Schema writerSchema = HoodieWriteHandle.createHoodieWriteSchema(config.getSchema());
```

Comment:
using the Handle classes at this level, breaks the layering.. Please refactor into a common helper.

Reply:
Moved to HoodieAvroUtils